### PR TITLE
Introduce APIs to register endHandler and exceptionHandler in Redis client (Pub/Sub)

### LINKED ISSUE #28696: Introduce APIs to register endHandler and exceptionHandler in Redis client (Pub/Sub)
### Description

Currently, there is no API to catch exceptions occurs at connection level and there is no API to handle onClose events, when using subscribe commands.

### Implementation ideas

The underlying client (vert.x Redis Client) exposes `RedisConnection#endHandler` and `RedisConnection#exceptionHandler` which are used to register callbacks.

### PR DESCRIPTION
**Pull request by Eng-Fouad**
Fixes #28696

### CODE REVIEW

**Comment by Eng-Fouad** [comment]
Alternative implementations:

- Add new methods in RedisSubscriber and ReactiveRedisSubscriber to register the callbacks.
- Add method that returns the redis connection object to have full control on it.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java [hunk 1]
```diff
@@ -25,45 +25,109 @@
...
      *
      * @param channel the channel
-     * @param onMessage the message consumer. Be aware that this callback is invoked for each message sent to the given
+     * @param onMessageHandler the message handler. Be aware that this callback is invoked for each message sent to the given
      *        channel, and is invoked on the <strong>I/O thread</strong>. So, you must not block. Offload to
      *        a separate thread if needed.
      * @return the subscriber object that lets you unsubscribe
      */
-    RedisSubscriber subscribe(String channel, Consumer<V> onMessage);
+    RedisSubscriber subscribe(String channel, Consumer<V> onMessageHandler);
```
Comment (cescoffier) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java:33]: Why renaming here? Handler is a vertx word.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java [hunk 2]
```diff
@@ -25,45 +25,109 @@
...
      *
      * @param pattern the pattern
-     * @param onMessage the message consumer. Be aware that this callback is invoked for each message sent to the
+     * @param onMessageHandler the message handler. Be aware that this callback is invoked for each message sent to the
      *        channels matching the pattern, and is invoked on the <strong>I/O thread</strong>. So, you must
      *        not block. Offload to a separate thread if needed.
      * @return the subscriber object that lets you unsubscribe
      */
-    RedisSubscriber subscribeToPattern(String pattern, Consumer<V> onMessage);
+    RedisSubscriber subscribeToPattern(String pattern, Consumer<V> onMessageHandler);
```
Comment (cescoffier) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java:44]: Same here

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java [hunk 3]
```diff
@@ -166,6 +195,24 @@ public Uni<String> subscribe() {
                         }
                     }
                 });
+                if (onEndHandler != null) {
+                    connection.endHandler(() -> {
+                        if (Vertx.currentContext() != null) {
+                            onEndHandler.run();
+                        } else {
+                            datasource.getVertx().runOnContext(onEndHandler);
```
Comment (cescoffier) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java:203]: Might need to create a duplicated context here

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java [hunk 4]
```diff
@@ -166,6 +195,24 @@ public Uni<String> subscribe() {
                         }
                     }
                 });
+                if (onEndHandler != null) {
+                    connection.endHandler(() -> {
+                        if (Vertx.currentContext() != null) {
+                            onEndHandler.run();
```
Comment (cescoffier) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java:201]: Need to check we are on a duplicated context and switch to one if needed

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java [hunk 5]
```diff
@@ -180,9 +227,9 @@ private void handleRedisEvent(UniEmitter<? super Void> emitter, Response r) {
             if ("subscribe".equalsIgnoreCase(command) || "psubscribe".equalsIgnoreCase(command)) {
                 emitter.complete(null); // Subscribed
             } else if ("message".equalsIgnoreCase(command)) {
-                context.runOnContext(x -> onMessage.accept(marshaller.decode(classOfMessage, r.get(2))));
+                context.runOnContext(x -> onMessageHandler.accept(marshaller.decode(classOfMessage, r.get(2))));
```
Comment (cescoffier) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java:230]: Note for myself: it must be a duplicated context. This need to be checked.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java (hunk 3 above)
Comment (Eng-Fouad) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java:203]: Something like this?

    connection.endHandler(() -> {
        Context ctxt = Vertx.currentContext();
        if (ctxt != null) {
            VertxContext.getOrCreateDuplicatedContext(ctxt)
                                 .runOnContext(ignored -> onEndHandler.run());
        } else {
            datasource.getVertx().runOnContext(() -> {
                VertxContext.getOrCreateDuplicatedContext(ctxt)
                                    .runOnContext(ignored -> onEndHandler.run());
            });
        }
    });

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java (hunk 1 above)
Comment (cescoffier) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/datasource/pubsub/PubSubCommands.java:33]: On would rename other methods too: onException, onEnd.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java (hunk 3 above)
Comment (cescoffier) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java]: The created duplicated context should also be marked as safe (or you won't be able to use hibernate reactive)

Comment (Eng-Fouad) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java]: How to mark it as safe?

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Can you rebase and squash your commits into a single one?

**Comment by Eng-Fouad** [comment]
> Can you rebase and squash your commits into a single one?

Done.

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java [hunk 6]
```diff
@@ -4,14 +4,12 @@
 import static io.smallrye.mutiny.helpers.ParameterValidation.doesNotContainNull;
 import static io.smallrye.mutiny.helpers.ParameterValidation.nonNull;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.List;
-import java.util.UUID;
+import java.util.*;
```
Comment (geoand) [inline extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java:7]: Please avoid star imports

**Review by geoand — COMMENTED** [review]
I added a trivial comment.

The rest is up to @cescoffier

File: extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java (hunk 6 above)
Comment (Eng-Fouad) [reply extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/ReactivePubSubCommandsImpl.java]: Fixed.

**Pull request merged**